test-results.cake

```diff
@@ -0,0 +1,182 @@
+
+		OverallResult = overallResult;
+		// Initialize counters to -1, indicating no expected value.
+		// Set properties of those items to be checked.
+		Total = Passed = Failed = Warnings = Inconclusive = Skipped = -1;
+	}
+}
+
+public class ActualResult : ResultSummary
+{
+    public ActualResult(string resultFile)
+    {
+        var doc = new XmlDocument();
+        doc.Load(resultFile);
+
+        Xml = doc.DocumentElement;
+        if (Xml.Name != "test-run")
+            throw new Exception("The test-run element was not found.");
+
+        OverallResult = GetAttribute(Xml, "result");
+        Total = IntAttribute(Xml, "total");
+        Passed = IntAttribute(Xml, "passed");
+        Failed = IntAttribute(Xml, "failed");
+        Warnings = IntAttribute(Xml, "warnings");
+        Inconclusive = IntAttribute(Xml, "inconclusive");
+        Skipped = IntAttribute(Xml, "skipped");
+    }
+
+    public XmlNode Xml { get; }
+
+    private string GetAttribute(XmlNode node, string name)
+    {
+        return node.Attributes[name]?.Value;
+    }
+
+    private int IntAttribute(XmlNode node, string name)
+    {
+        string s = GetAttribute(node, name);
+        return s == null ? 0 : int.Parse(s);
+    }
+}
+
+public class TestReport
+{
+    public PackageTest Test;
+    public ActualResult Result;
+    public List<string> Errors;
+
+    public TestReport(PackageTest test, ActualResult result)
+    {
+        Test = test;
+        Result = result;
+        Errors = new List<string>();
+
+        var expected = test.ExpectedResult;
+
+        ReportMissingFiles();
+
+        if (result.OverallResult == null)
+            Errors.Add("     The test-run element has no result attribute.");
+        else if (expected.OverallResult != result.OverallResult)
+            Errors.Add($"     Expected: Overall Result = {expected.OverallResult}\n    But was: {result.OverallResult}");
+        CheckCounter("Test Count", expected.Total, result.Total);
+        CheckCounter("Passed", expected.Passed, result.Passed);
+        CheckCounter("Failed", expected.Failed, result.Failed);
+        CheckCounter("Warnings", expected.Warnings, result.Warnings);
+        CheckCounter("Inconclusive", expected.Inconclusive, result.Inconclusive);
+        CheckCounter("Skipped", expected.Skipped, result.Skipped);
+    }
+
+    public TestReport(PackageTest test, Exception ex)
+    {
+        Test = test;
+        Result = null;
+        Errors = new List<string>();
+        Errors.Add($"     {ex.Message}");
+    }
+
+    public void Display(int index)
+    {
+        Console.WriteLine($"\n{index}. {Test.Description}");
+        Console.WriteLine($"   Args: {Test.Arguments}\n");
+
+        foreach (var error in Errors)
+            Console.WriteLine(error);
+
+        Console.WriteLine(Errors.Count == 0
+            ? "   SUCCESS: Test Result matches expected result!"
+            : "\n   ERROR: Test Result not as expected!");
+    }
+
+    private void ReportMissingFiles()
```

Comment:
I'm a little confused by what this method's doing sorry. It looks to me like it's recording errors for all the failed test suites, but if I'm right about that (I suspect I'm not!), then I don't understand why it's named `ReportMissingFiles`. Do you think it might be worth commenting this method?

Reply:
What it looks at is not all test suites, but all those that may occur at or above the assembly level. Project and Assembly are self-explanatory, while Unknown is what we get if the file is either not found or is of an unknown type. (BTW I've tested all these cases manually or I wouldn't have known that myself.) Within those suites, we only look at those that failed __themselves__ (not due to a child test failing) and display the error message.

Without the errors, you can't figure out why all your actuals are zero. With it you get to see messages like "File not found", "Not a supported extension", "Invalid XML", etc. without scrolling back to look at the actual test run or in some cases running the debugger. It saved me a lot of time working on this.

I'll write all that in a comment.

Reply:
Great, thanks! 😄 Makes complete sense.